Add ref/out keywords on arguments for shadow methods in GeneratedComInterface

jkoritzinsky opened the pull request:
Fix a regression blocking https://github.com/dotnet/sdk/pull/43015

jkoritzinsky committed:
Add ref kinds on arguments for shadow methods

jtschuster reviewed (approved):
nit: I don't think the .sln changes are necessary.

AaronRobinsonMSFT commented:
Is this a regression? If so, what did we miss?

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/CodeSnippets.cs:
```diff
@@ -671,6 +671,36 @@ interface IOtherInterface
+            {
+            }
+
+            {{GeneratedComInterface()}}
+            partial interface INativeAPIBase
+            {
+                [PreserveSig]
+                int FindValue(int key, out IValue? value);
+            }
+
+            {{GeneratedComInterface()}}
+            unsafe partial interface INativeDerived : INativeAPIBase
+            {
+                [PreserveSig]
+                int GetName(out char* name);
+            }
```
Do we already have a test for `ref`? What about `in`?

AaronRobinsonMSFT replied:
If not, feel free to add them to the existing signature. or a new one for kind.

Also, `ref readonly`?

jkoritzinsky replied:
I'll refactor this test to validate all of the kinds.

jkoritzinsky commented:
Yeah this was a regression from one of the refactorings I did when I simplified the bound generator concept in the source generator APIs (and moved this part to not use that infrastructure)

jkoritzinsky committed:
Revert sln changes and expand test.

jkoritzinsky commented:
Looks like we have one more bug in here. I'll push out a fix.

jkoritzinsky committed:
Use in for ref readonly to support forwarding ref readonly parameters
